### ISSUE TITLE
Reference observation doesn't listen in on updated subtrees

### ISSUE THREAD

**Issue by Dashed** [omniscientjs/immstruct#37]
# tl;dr

```js
let structure = immstruct();

// this doesn't work as intended. very unexpected when user doesn't know how
// Cursor.from(...) works.
structure
.reference(['config', 'profile'])
.observe(function(newObj, oldObj, hotKeyPath) {
    // no output
    console.log(arguments);
});


// update some object
structure.cursor().set('config', Immutable.fromJS({
    'profile': 'throwaway',
    'color': 'green'
}));
```

-----

# working solution

```js
let immstruct = require('immstruct');
let Immutable = require('immutable');

let structure = immstruct();

// this doesn't work as intended. very unexpected when user doesn't know how
// Cursor.from(...) works.
structure
.reference(['config', 'profile'])
.observe(function(newObj, oldObj, hotKeyPath) {
    // no output
    console.log(arguments);
});

const NOT_SET = {};

const rootRef = structure.reference();

// much more useful observation...
function betterObserve(observeKeyPath, eventName, newFn) {

    // this probably calls for self._pathListeners to be a tree structure or
    // something.

    return rootRef.observe(function(newRoot, oldRoot, hotKeyPath) {

        if(newRoot === oldRoot) {
            return;
        }

        let
        idx = 0,
        newObj = newRoot,
        oldObj = oldRoot,
        hotKeyPathLen = hotKeyPath.length;

        for(let len = observeKeyPath.length; idx < len; idx++) {

            let atomPath = observeKeyPath[idx];

            // verify if hotKeyPath is a subpath of observeKeyPath
            if(idx < hotKeyPathLen && atomPath !== hotKeyPath[idx]) {
                return;
            }

            if((newObj == NOT_SET && oldObj !== NOT_SET)
                || (oldObj == NOT_SET && newObj !== NOT_SET)) {
                continue;
            }

            newObj = newObj.get(atomPath, NOT_SET);
            oldObj = oldObj.get(atomPath, NOT_SET);

            if(newObj === oldObj) {
                return;
            }
        }

        if (eventName && eventName !== 'swap') {
            let info = analyze(newRoot, oldRoot, observeKeyPath);
            if (info.eventName !== eventName) return;
            return newFn.apply(newFn, info.arguments);
        }

        return newFn(newRoot, oldRoot, observeKeyPath);
    });

};

betterObserve(['config', 'profile'], 'add', function() {
    // outputs: { '0': [ 'config', 'profile' ], '1': 'default' }
    console.log(arguments);
});

// update some object
structure.cursor().set('config', Immutable.fromJS({
    'profile': 'default',
    'color': 'green'
}));
````

**Comment by Dashed** [comment]
We cannot rely on changed `keyPath` emitted from `Cursor.from(...)`, because any sub-trees at `keyPath` may have changed. Hence, signals would be lost.

**Comment by Dashed** [comment]
I looked a bit into generalized suffix trees; I don't think we'd need to implement it without having to add so much weight to immstruct.

**Comment by Dashed** [comment]
Link to gitter comments for posterity:

https://gitter.im/omniscientjs/omniscient?at=54f4b2e7b1086da6119df8e5
https://gitter.im/omniscientjs/omniscient?at=54f4b43cdf52f34314e4f974

**Comment by jabjab** [comment]
This doesn't just affect `.observe`. (Or, perhaps, the effect is due to that function being affected.)

```javascript
let structure = immstruct();
let ref = structure.reference(['a', 'b']);

structure.cursor().set('a', Immutable.Map({b: 3}));

console.log(ref.cursor().deref()); // undefined
```

**Comment by Dashed** [comment]
@jabjab Thanks for pointing out that case.

Unfortunately, doing `structure.cursor()` sets `keypath = []`. So the listening reference won't realize the change: https://github.com/omniscientjs/immstruct/blob/7c9db6ccd263d102a0f7d07ade3fffa360af7d1f/src/structure.js#L65

-----

IMO, `Cursor.from(...)` is the wrong way to observe subtrees. So I'm currently experimenting state-less cursors with first-class observation. 

An idea I'm playing with is making `Structure` a state-less cursor type that will have proper subtree observation. Still sketching out the API though.

**Comment by andrewluetgers** [comment]
@Dashed I totally agree that sub paths need to trigger change events, I also ran into this problem and it was kind of earth shattering to how I envisioned our app working. It adds and annoying layer of complexity to state-management that really needs to go away. As you rightly point out a tree probably is the correct structure as you have noticed on my initial hack at that idea I was able to achieve some good results.
